### PR TITLE
Add support for judge/normal ratio in voting system

### PR DESCRIPTION
**Pull request by RaidoAun**
In admin panel the ratio between normal votes and judge votes can be set on and event. Also when approving jammers you can now set them as judges.

### CODE REVIEW

File: datastoredb/events.go [hunk 1]
```diff
@@ -473,8 +473,11 @@ func (repo *Events) Results(eventid event.EventID) ([]*event.TeamResult, error)
 	}
 
 	results := createTeamResults(teams, ballots)
+
+	currentEvent, err := repo.ByID(eventid)
```
Comment (egonelbre) [inline datastoredb/events.go:477]: `err` was not handled here

File: event/admin.go [hunk 2]
```diff
@@ -124,20 +126,28 @@ func (server *Server) EditEvent(context *Context) {
 		}
 
 		theme := context.FormValue("theme")
+		
+		judgePercentage, err := strconv.ParseFloat(context.FormValue("judgePercentage"),64)
```
Comment (egonelbre) [inline event/admin.go:130]: handle this based on whether it has been specified or not

File: event/admin.go [hunk 3]
```diff
@@ -124,20 +126,28 @@ func (server *Server) EditEvent(context *Context) {
...
+		if err != nil {
+			judgePercentage = 50.0
+			log.Println(err.Error())
```
Comment (egonelbre) [inline event/admin.go:133]: return the error -- there shouldn't be a case where we try to parse and it fails and we just log it

most likely you want to check whether the value has been specified and then either set it or not

File: event/admin.go [hunk 4]
```diff
@@ -124,20 +126,28 @@ func (server *Server) EditEvent(context *Context) {
...
+		}
+		
 		registration := context.FormValue("registration") == "true"
 		voting := context.FormValue("voting") == "true"
 		closed := context.FormValue("closed") == "true"
 		revealed := context.FormValue("revealed") == "true"
 		info := context.FormValue("info")
-
+
```
Comment (egonelbre) [inline event/admin.go:141]: There are tabs here that shouldn't exist. Format code with `go fmt`

File: event/admin.go [hunk 5]
```diff
@@ -261,12 +285,19 @@ func (server *Server) Jammers(context *Context) {
 		}
 
 		s := ""
-		if len(removed) > 0 {
-			s += fmt.Sprintf(" Removed %v jammers.", len(removed))
+		if len(jammersRemoved) > 0 {
+			s += fmt.Sprintf(" Removed %v jammers.\n", len(jammersRemoved))
```
Comment (egonelbre) [inline event/admin.go:289]: This formatting doesn't look right in the case only some of the branches are taken. You want to ensure that there's no trailing whitespace. One way to achieve this is to use:

```
flashes := []string{}
if len(jammersRemoved) > 0 {
			flashes = append(flashes, fmt.Sprintf("Removed %v jammers.", len(jammersRemoved)))
...
s := strings.Join(flashes, "\n")
```

File: event/ballot.go [hunk 6]
```diff
@@ -172,6 +200,15 @@ func (aspects *Aspects) Add(other *Aspects) {
 	aspects.Overall.Score += other.Overall.Score
 }
 
+func (aspects *Aspects) MultiplyByScalar(multiplier float64) {
```
Comment (egonelbre) [inline event/ballot.go:203]: Just `Scale` as the method name would be sufficient.

File: templates/event/event-create.html [hunk 7]
```diff
@@ -18,6 +18,12 @@ <h1>Create Event</h1>
 			<input type="text" id="theme" name="theme" value="{{.Event.Theme}}">
 		</div>
 
+		<div class="field">
+			<label for="judgePercentage">Percentage of judge votes</label>
+			<input type="range" min="0" max="100" step="10" id="judgePercentage" name="judgePercentage" value="{{.Event.JudgePercentage}}" oninput="(()=>{document.getElementById('judgePercentageInfoText').innerText = `Current percentage: ${document.getElementById('judgePercentage').value}%`})()">
```
Comment (egonelbre) [inline templates/event/event-create.html:23]: A number field might be nicer to use in this case. For events you usually want to specify a concrete value.

File: event/ballot.go [hunk 8]
```diff
@@ -67,32 +67,60 @@ var DefaultAspects = Aspects{
...
+	judgeCount := 0.0
+	normalSum := Aspects{}
+	judgeSum := Aspects{}
+
 	for _, ballot := range ballots {
 		if !ballot.Completed {
 			continue
 		}
 
-		if count == 0 {
```
Comment (egonelbre) [inline event/ballot.go:78]: I cannot recall, but I'm not sure whether this case is even necessary.

File: event/ballot.go [hunk 9]
```diff
@@ -67,32 +67,60 @@ var DefaultAspects = Aspects{
...
 // AverageScores returns averages for all aspects.
-func AverageScores(ballots []*Ballot) Aspects {
-	count := 0.0
-	average := Aspects{}
+func AverageScores(ballots []*Ballot, event *Event) Aspects {
+	judgePercentage := event.JudgePercentage
+
+	normalCount := 0.0
+	judgeCount := 0.0
+	normalSum := Aspects{}
```
Comment (egonelbre) [inline event/ballot.go:75]: just `jammers`

File: event/ballot.go [hunk 10]
```diff
@@ -67,32 +67,60 @@ var DefaultAspects = Aspects{
...
 	}
 
-	if count > 0 {
-		average.Theme.Score /= count
-		average.Enjoyment.Score /= count
-		average.Aesthetics.Score /= count
-		average.Innovation.Score /= count
-		average.Bonus.Score /= count
-		average.Overall.Score /= count
+	if judgePercentage != 100 {
```
Comment (egonelbre) [inline event/ballot.go:101]: Create a new variable for the average.

It also looks like the calculation could be simplified, but I need to think how to write it nicer.

Most likely something like this (ignoring mistakes):
```
if jammerCount > 0 { jammers.Scale(1 / jammerCount) }
if judgeCount > 0 { judges.Scale(1 / judgeCount) }

jammers.Scale(1 - judgePercentage/100)
judges.Scale(judgePercentage/100)

var final Aspects
final.Add(jammers)
final.Add(judges)
```

File: event/admin.go [hunk 11]
```diff
@@ -124,20 +126,28 @@ func (server *Server) EditEvent(context *Context) {
 		}
 
 		theme := context.FormValue("theme")
+		
+		judgePercentage, err := strconv.ParseFloat(context.FormValue("judgePercentage"),64)
+		if err != nil {
+			judgePercentage = 50.0
```
Comment (egonelbre) [inline event/admin.go:132]: Let's default to this to 0 or undefined, as this is the case for all previous events.

File: event/ballot.go [hunk 12]
```diff
@@ -67,32 +67,60 @@ var DefaultAspects = Aspects{
...
-func AverageScores(ballots []*Ballot) Aspects {
-	count := 0.0
-	average := Aspects{}
+func AverageScores(ballots []*Ballot, event *Event) Aspects {
+	judgePercentage := event.JudgePercentage
+
+	normalCount := 0.0
+	judgeCount := 0.0
+	normalSum := Aspects{}
+	judgeSum := Aspects{}
```
Comment (egonelbre) [inline event/ballot.go:76]: just `judges`

File: event/ballot.go [hunk 13]
```diff
@@ -67,32 +67,60 @@ var DefaultAspects = Aspects{
 }
 
 // AverageScores returns averages for all aspects.
-func AverageScores(ballots []*Ballot) Aspects {
-	count := 0.0
-	average := Aspects{}
+func AverageScores(ballots []*Ballot, event *Event) Aspects {
```
Comment (egonelbre) [inline event/ballot.go:70]: We also probably want to show the "judges" and "jammer" scores separately.

If you separate ballots by whether a user is a judge and whether not, then the code might become even more clearer, e.g.


```
jammer := AverageScores(jammerBallots)
judges := AverageScores(judgeBallots)
```

Then you can calculate the average much nicer. Of course, there's a bit more overhead.

**Comment by egonelbre** [comment]
Forgot to mention, the idea seems ok, and good work.

File: event/ballot.go (hunk 8 above)
Comment (RaidoAun) [reply event/ballot.go:78]: All seems to work without it. Removed it.

File: event/ballot.go (hunk 13 above)
Comment (RaidoAun) [reply event/ballot.go]: Doing it this way would have forced me to duplicate some of the AverageScores logic when calculating the total average.
Now AverageScores just returns them all seperately.

**Review by egonelbre — APPROVED** [review]

**Pull request merged**